### PR TITLE
Add a doc about end-to-end tensor expressions workflow.

### PR DESCRIPTION
**Pull request by ZolotukhinM**

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/DesignOverview.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+## Tensor
+`Tensor` is a bundle of
+1) a variable `Var` defining which tensor this `Tensor` expression is describing
+2) a list of indices `args` (each of them is `Var`)
+3) a list of expressions for dimensions `dims` (each of them is `Expr`)
+4) a computational expression `body` (of `Expr` type)
+
+## Buffer
+`Buffer`s are essentially `Tensor`s without a `body` - they represent an indexed access to something not defined in the tensor-expression.
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/DesignOverview.md]: Minor: They represent a symbolic access to "tensors" that is outside the tensor-expression system.

**Review by bertmaher — APPROVED** [review]
This is a nice looking doc!

**Review by bertmaher — COMMENTED** [review]
Do we usually do 80col for markdown docs?  I like 80col but it's just because I always have like 4 emacs buffers spanning my monitor :)

File: torch/csrc/jit/tensorexpr/DesignOverview.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+1) a `Var` defining which buffer this `Buffer` expression is defining
+2) a list of indices `args` (each of them is `Var`)
+3) a list of expressions for dimensions `dims` (each of them is `Expr`)
+
+## Example
+Suppose we'd like to represent the following expression:
+```
+A[i,j] = B[i,j] + 7
+```
+where both `A` and `B` are 100x100 tensors.
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/DesignOverview.md:78]: Minor: In this example, include a section using "Compute".

**Review by zheng-xq — APPROVED** [review]
Looks great! Some minor comments.

File: torch/csrc/jit/tensorexpr/DesignOverview.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+## Step 4: Lower the tree structure to statements.
+This step creates a `For` statement for each loop-axis and emits `element_stmt` for bodies of the loops.
+
+## Step 5: Pass the final statement for codegen (LLVM/CUDA/IREval)
+Codegen is implemented as an IR visitor over the statements produced in the previous step.
+
+# Tensor Expressions Language
+There are several core concepts in the Tensor Expression engine, this section defines them and shows how they connect to each other.
+
+## Expr
```
Comment (protonu) [inline torch/csrc/jit/tensorexpr/DesignOverview.md:56]: Thanks! This is helpful. I was wondering if you could add a few lines for Stmt as well

File: torch/csrc/jit/tensorexpr/DesignOverview.md (hunk 1 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/DesignOverview.md]: Right, that's what I tried to say, I'll borrow your version :)

File: torch/csrc/jit/tensorexpr/DesignOverview.md (hunk 3 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/DesignOverview.md:56]: Oh, right, I forgot about them. However, I think `Stmt`s should not be considered part of the TE language (even though currently they look like they are) - if we manage to finish the Expr/Stmt separation, the statements would never come up at the stage of constructing a tensor expression and would only appear as an implementation detail of how we lower the expressions into loop-nests (currently: schedules)

**Comment by ZolotukhinM** [comment]
> Do we usually do 80col for markdown docs? I like 80col but it's just because I always have like 4 emacs buffers spanning my monitor :)

I actually specifically turned off 80col for this doc - this way it renders correctly in a web-preview without random line breaks.

**Pull request merged**